packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol

```diff
@@ -39,44 +42,64 @@ contract ReserveCurrencyLiquidator {
-        FixedPoint.Unsigned memory totalCollateralNeeded =
-            collateralToMintAtGcr.add(store.computeFinalFee(fc.collateralCurrency()));
-
-        // 3. Swap reserve currency to get required collateral for the mint + final fee.
-        address[] memory path = new address[](2);
-        path[0] = reserveCurrency;
-        path[1] = fc.collateralCurrency();
-
-        TransferHelper.safeApprove(reserveCurrency, address(router), maxReserverTokenSpent.rawValue);
-        router.swapTokensForExactTokens(
-            totalCollateralNeeded.rawValue,
-            maxReserverTokenSpent.rawValue,
-            path,
-            address(this),
-            deadline
-        );
+        // 1. Calculate the token shortfall. This is the synthetics to liquidate minus any synthetics the DSProxy already
+        // has. If this number is negative(balance large than synthetics to liquidate) the return 0 (no shortfall).
+        FixedPoint.Unsigned memory tokenShortfall =
+            subOrZero(maxTokensToLiquidate, FixedPoint.Unsigned(IERC20(fc.tokenCurrency()).balanceOf(address(this))));
 
-        // 4. Mint synthetics with collateral. Note we are minting at the GCR and minting the exact number liquidated.
-        TransferHelper.safeApprove(fc.collateralCurrency(), address(fc), totalCollateralNeeded.rawValue);
-        fc.create(collateralToMintAtGcr, maxTokensToLiquidate);
-
-        // 5. Liquidate position with newly minted synthetics.
+        // 2. Calculate how much collateral is needed to make up the token shortfall from minting new synthetics.
+        FixedPoint.Unsigned memory gcr = fc.pfc().div(fc.totalTokensOutstanding());
+        FixedPoint.Unsigned memory collateralToMintShortfall = tokenShortfall.mul(gcr);
+
+        // 3. Calculate the total collateral required. This considers the final fee for the given collateral type + any
+        // collateral needed to mint the token short fall.
+        FixedPoint.Unsigned memory totalCollateralRequired =
+            IStore(IFinder(fc.finder()).getImplementationAddress("Store")).computeFinalFee(fc.collateralCurrency()).add(
+                collateralToMintShortfall
+            );
+
+        // 4. Calculate how much collateral needs to be purchased. If the DSProxy already has some collateral then this
+        // will factor this in. If the DSProxy has more collateral than the total amount required the purchased = 0.
+        FixedPoint.Unsigned memory collateralToBePurchased =
+            subOrZero(
+                totalCollateralRequired,
+                FixedPoint.Unsigned(IERC20(fc.collateralCurrency()).balanceOf(address(this)))
+            );
+
+        // 4.a. If there is some collateral to be purchased, execute a trade on uniswap to meet the shortfall.
+        // Note the path assumes a direct route from the reserve currency to the collateral currency.
+        if (collateralToBePurchased.isGreaterThan(FixedPoint.fromUnscaledUint(0))) {
```

Comment:
What if `fc.collateralCurrency() == reserveCurrency`?

Reply:
This is a great question. I'm going to write some unit tests to verify this but my understanding is that the trade will just not happen. might be better to have some branching logic

Reply:
I added a test which validates this. Simply no trade happens and everything else continues on as it was before. no additional branching is needed; the `swapTokenForExactTokens` simply does nothing and returns :)